src/dune_rules/gen_rules.ml

```diff
@@ -328,105 +328,120 @@ let gen_project_rules sctx project =
-                Memo.Build.return ()
-            in
-            (* This interprets "rule" and "copy_files" stanzas. *)
-            Dir_contents.gen_rules sctx ~dir >>= function
-            | Group_part root ->
-              Load_rules.load_dir_and_produce_its_rules ~dir:(Path.build root)
-            | Standalone_or_root (dir_contents, subs) ->
-              let* cctxs = gen_rules sctx dir_contents [] ~source_dir ~dir in
-              Memo.Build.parallel_iter subs ~f:(fun dc ->
-                  gen_rules sctx dir_contents cctxs ~source_dir
-                    ~dir:(Dir_contents.dir dc)
-                  >>| ignore)))
-      in
-      S.These (String.Set.of_list subdirs)
-  in
-  let subdirs_to_keep2 =
-    match components with
-    | [] ->
-      Subdir_set.These (String.Set.of_list [ ".js"; "_doc"; ".ppx"; ".dune" ])
-    | _ -> These String.Set.empty
-  in
-  Subdir_set.union_all [ subdirs_to_keep1; subdirs_to_keep2 ]
+  match components with
+  | [ ".dune"; "ccomp" ] ->
+    has_rules
+      ((* Add rules for C compiler detection *)
+       let+ () = Cxx_rules.rules ~sctx ~dir in
+       S.These String.Set.empty)
+  | ".dune" :: _ -> has_rules (Memo.Build.return (S.These String.Set.empty))
+  | ".js" :: rest ->
+    has_rules
+      (let+ () = Jsoo_rules.setup_separate_compilation_rules sctx rest in
+       match rest with
+       | [] -> S.All
+       | _ -> S.These String.Set.empty)
+  | "_doc" :: rest ->
+    has_rules
+      (let+ () = Odoc.gen_rules sctx rest ~dir in
+       match rest with
+       | [] -> S.All
+       | _ -> S.These String.Set.empty)
+  | ".ppx" :: rest ->
+    has_rules
+      (let+ () = Preprocessing.gen_rules sctx rest in
+       match rest with
+       | [] -> S.All
+       | _ -> S.These String.Set.empty)
+  | comps -> (
+    Source_tree.find_dir (Path.Build.drop_build_context_exn dir) >>= function
+    | None -> (
+      let has_rules x = has_rules (x >>> Memo.Build.return Subdir_set.empty) in
+      match List.last comps with
+      | Some ".formatted" ->
+        has_rules
+          (let* expander = Super_context.expander sctx ~dir in
+           gen_format_rules sctx ~expander ~output_dir:dir)
+      | Some ".bin" ->
+        has_rules
+          (let src_dir = Path.Build.parent_exn dir in
+           let* local_binaries =
+             Super_context.local_binaries sctx ~dir:src_dir
+           in
+           Memo.Build.sequential_iter local_binaries ~f:(fun t ->
+               let loc = File_binding.Expanded.src_loc t in
+               let src = Path.build (File_binding.Expanded.src t) in
+               let dst = File_binding.Expanded.dst_path t ~dir in
+               Super_context.add_rule sctx ~loc ~dir
+                 (Action_builder.symlink ~src ~dst)))
+      | Some ".utop" ->
+        has_rules (Utop.setup sctx ~dir:(Path.Build.parent_exn dir))
+      | Some _ -> redirect_to_parent
+      | None ->
+        (* [Source_tree.find_dir] never returns [None] for the root, so we
```

Comment:
I'm confused: this `None` seems to be the result of `List.last comps`, not `Source_tree.find_dir`.

Reply:
It's because the two are tied. But yeah, maybe this comment is just confusing. I can remove it (it's not present in the final state BTW)